apps/src/dance/ai/AiBlockPreview.tsx

```diff
@@ -43,23 +40,25 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 
   // Build out the blocks.
   useEffect(() => {
-    if (!done) {
-      const emptyBlockXml = Blockly.utils.xml.textToDom('<xml></xml>');
-      const previewWorkspace: Workspace =
-        Blockly.BlockSpace.createReadOnlyBlockSpace(
-          blockPreviewContainerRef.current,
-          emptyBlockXml,
-          {}
-        );
+    const emptyBlockXml = Blockly.utils.xml.textToDom('<xml></xml>');
+    const previewWorkspace: Workspace =
```

Comment:
Thanks for doing this! I was looking at this last week and wondering if we should only create the workspace once (maybe in a separate useEffect that only responds to changes in the container ref)?

Reply:
Ah good question - though I think this allows us to be more strict about the workspace lifecycle, since we don't want to be rendering additional blocks even if we have a single workspace? This way the workspace is only created once (and shouldn't be recreated unnecessarily due to the `useCallback` wrapping `generateBlocksFromResult`) and we can also ensure that we call `workspace.dispose()` when unmounting, or when dependencies do happen to change.

Reply:
Ok that makes sense -- I guess I was wondering what happens if this were to get called multiple times (acknowledging with the changes you've made to `generateBlocksFromResult` that it shouldn't), like what happens if a second workspace is generated within the same ref? Anyway probably doesn't matter

Reply:
Yep yep I think still in that case we'd be okay since we clean up with `workspace.dispose()` - so if the `useEffect` ever runs multiple times, we'll call `dispose()` before each run which should remove any existing blocks before rendering the new workspace

Reply:
Oh snap, I think I misunderstood react docs -- the cleanup callback from useEffect runs not only on unmount but [before the subsequent render](https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount). Nice!

Reply:
Thanks for the discussion and shared reference! TIL about cleanup functions in `useEffect`. 😃